lib/ansible/modules/cloud/azure/azure_rm_image.py

```diff
@@ -0,0 +1,368 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2017 Yuwei Zhou, <yuwzho@microsoft.com>
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'certified'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_image
+version_added: "2.5"
+short_description: Manage Azure image.
+description:
+    - Create, delete an image from virtual machine, blob uri, managed disk or snapshot.
+options:
+    resource_group:
+        description:
+            - Name of resource group.
+        required: true
+    name:
+        description:
+            - Name of the image.
+        required: true
+    source:
+        description:
+            - OS disk source from the same region, including a virtual machine id or name,
+              OS disk blob uri, managed OS disk id or name, or OS snapshot id or name.
+        required: true
+    data_disk_sources:
+        description:
+            - List of data disk sources, including unmanaged blob uri, managed disk id or name, or snapshot id or name.
+    location:
+        description:
+            - Location of the image.
+    os_type:
+        description: The OS type of image.
+        choices:
+            - Windows
+            - Linux
+    state:
+        description:
+            - Assert the state of the image. Use 'present' to create or update a image and 'absent' to delete an image.
+        default: present
+        choices:
+            - absent
+            - present
+
+extends_documentation_fragment:
+    - azure
+    - azure_tags
+
+author:
+    - "Yuwei Zhou (@yuwzho)"
+
+'''
+
+EXAMPLES = '''
+- name: Create an image from a virtual machine
+  azure_rm_image:
+    resource_group: Testing
+    name: foobar
+    source: testvm001
+
+- name: Create an image from os disk
+  azure_rm_image:
+    resource_group: Testing
+    name: foobar
+    source: /subscriptions/XXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXX/resourceGroups/Testing/providers/Microsoft.Compute/disks/disk001
+    data_disk_sources:
+        - datadisk001
+        - datadisk002
+    os_type: Linux
+
+- name: Delete an image
+  azure_rm_image:
+    state: absent
+    resource_group: Testing
+    name: foobar
+    source: testvm001
+'''
+
+RETURN = '''
+state:
```

Comment:
Can we filter the values we return to only be things we don't know.

* The ID of the disk
* Potentially the underlying data disks of the image

These values should also be set on the root level and not in `state` so I could do

```
- name: create image
  azure_rm_image:
    resource_group: Testing
    name: foobar
    source: testvm001

  register: image_info

- name: show image id
  debug:
    var: image_info.id
```

The facts module is a different story but we usually only return values for things we don't know.

Reply:
I m confused what should we return in the module. Is there any overhead for return everything I know to user? Consider the maintenance, it seems more convenient if we keep the struct as Azure SDK returned instead of flatting it and deleting some information in the module layer.

Reply:
There are a few reasons;

* if you are just converting the result from the dict  you are bound the following restrictions
    * any changes in this return value get's reflected to the module and if a playbook is reliant on these values they will now break, we want to keep things consistent
    * the return values are akin to being an API as well, we want to ensure we keep things as simple so that we don't accrue technical debt
    * the structure itself can be in camelCase format which we try to avoid in Ansible to keep consistency in the values
    * Sometimes what people look for is deeply nested in the return value and should be more accessible, case in point when trying to get the public IP for an instance created with `azure_rm_virtualmachine` I need to do `result.properties.networkProfile.networkInterfaces[0].properties.ipConfigurations[0].properties.publicIPAddress.properties.ipAddress`. Compare this to the EC2 module, to get the public IP I only need to do `result.instances[0].public_ip` which is a hell of a lot easier for a non programmer to derive and understand
* we usually advocate only sending info we don't know already (in this case the ID) as the return values are mostly used for further actions on the newly created object. If we wanted to get more information about an object, that's why we have facts modules.
* each byte being sent across the wire takes more time, while in this case it would make a marginal difference why would you need to send the name of an image when you already know that value

Ultimately the first point is the main reason, and it comes down to the fact that we want a nice and consistent interface that doesn't change over time. In a perfect world we could probably have that but unfortunately that is not this world so we keep things simple as we can.

Reply:
Thanks for you explain, I have changed the _image_to_dict from simply serialize it to flatten it.

Reply:
No worries, I see your changes but it just needs to be taken 1 step further. Right now the results are being added to `result['state']` when each value should just be in result, e.g. ID should be `result['id']` and not `result['state']['id']`

Reply:
I see your example that some modules return properties in the result's child level. But all other azure modules return property values in the state level. It is so strange that the azure image module return thing not consistent with other azure modules. A user who uses Azure modules may confuse with this.